### PR TITLE
Feature/feature better python 3

### PR DESCRIPTION
**Pull request by CFlaniganMide**
Re-did the python3 conversion for ebmlite from the ground up with future packages and tools and it just went infinitely easier.

### CODE REVIEW

**Comment by CFlaniganMide** [comment]
Needs more work:
- Infinite elements still don't seem to be handled correctly.  
- Dependency on future needs to be removed.

File: ebmlite/core.py [hunk 1]
```diff
@@ -708,19 +697,19 @@ def gc(self, recurse=False):
     def encodePayload(cls, data, length=None):
         """ Type-specific payload encoder for 'master' elements.
         """
-        result = u''
+        result = bytearray()
         if data is None:
             return result
         elif isinstance(data, dict):
-            data = data.items()
+            data = list(data.items())
```
Comment (StokesMIDE) [inline ebmlite/core.py:704]: Casting to list does avoid the risk of the dict changing during encoding, but is it worth the memory usage? These could be *massive* elements. `data.items()` should work as-is in the loop.

File: ebmlite/core.py [hunk 2]
```diff
@@ -35,36 +35,34 @@
...
-from __future__ import division, absolute_import, print_function, unicode_literals
+from __future__ import absolute_import, division, print_function, unicode_literals
 
-__author__ = "David Randall Stokes"
-__copyright__ = "Copyright 2018 Mide Technology Corporation"
+from future import standard_library
+standard_library.install_aliases()
+from builtins import bytes
+from past.builtins import basestring
+from builtins import object
```
Comment (StokesMIDE) [inline ebmlite/core.py:44]: Do you need to import anything from `builtins`?

File: ebmlite/core.py [hunk 3]
```diff
@@ -35,36 +35,34 @@
...
+__author__ = b"David Randall Stokes"
```
Comment (StokesMIDE) [inline ebmlite/core.py:45]: Metadata is supposed to go after `__future__` imports but before regular imports (PEP8).

File: ebmlite/core.py [hunk 4]
```diff
@@ -244,11 +241,11 @@ def encode(cls, value, length=None, lengthSize=None, infinite=False):
             @return: A bytearray containing the encoded EBML data.
         """
         if infinite and not issubclass(cls, MasterElement):
-            raise ValueError("Only Master elements can have 'infinite' lengths")
+            raise ValueError(b"Only Master elements can have 'infinite' lengths")
```
Comment (StokesMIDE) [inline ebmlite/core.py:244]: Why are these byte strings? They're for human consumption.

File: ebmlite/core.py [hunk 5]
```diff
@@ -106,7 +104,7 @@ class Element(object):
         @cvar length: An explicit length (in bytes) of the element when
             encoding. `None` will use standard EBML variable-length encoding.
     """
-    __slots__ = ("stream", "offset", "size", "sizeLength", "payloadOffset", "_value")
+    __slots__ = ("stream", "offset", "size", "payloadOffset", "_value")
```
Comment (StokesMIDE) [inline ebmlite/core.py:107]: Was `sizeLength` unused?

File: ebmlite/core.py [hunk 6]
```diff
@@ -616,7 +605,7 @@ def size(self):
                         break
                 except TypeError as err:
                     # Will occur at end of file; message will contain "ord()".
-                    if "ord()" in str(err):
+                    if b"ord()" in bytes(err):
```
Comment (StokesMIDE) [inline ebmlite/core.py:608]: I don't think this is necessary.

File: ebmlite/core.py [hunk 7]
```diff
@@ -826,21 +815,26 @@ def __init__(self, stream, name=None, size=None, headers=True):
                 self.name = self.__class__.__name__
 
         if size is None:
-            # Note: this doesn't work for cStringIO!
-            if isinstance(stream, StringIO):
+            # Note: this doesn't work for cBytesIO!
```
Comment (StokesMIDE) [inline ebmlite/core.py:818]: is `cBytesIO` a thing?

File: ebmlite/core.py [hunk 8]
```diff
@@ -904,21 +898,21 @@ def __getitem__(self, idx):
         """ Get one of the document's root elements by index.
         """
         # TODO: Cache parsed root elements, handle indexing dynamically.
-        if isinstance(idx, (int, long)):
+        if isinstance(idx, (int, int)):
```
Comment (StokesMIDE) [inline ebmlite/core.py:901]: Redundant `int`

File: ebmlite/core.py [hunk 9]
```diff
@@ -904,21 +898,21 @@ def __getitem__(self, idx):
...
             if idx < 0:
-                raise IndexError("Negative indices in a Document not (yet) supported")
+                raise IndexError(b"Negative indices in a Document not (yet) supported")
```
Comment (StokesMIDE) [inline ebmlite/core.py:903]: I'm not sure any exception should have bytestring messages.

File: ebmlite/encoding.py [hunk 10]
```diff
@@ -4,23 +4,21 @@
...
-__author__ = "dstokes"
-__copyright__ = "Copyright 2018 Mide Technology Corporation"
+from builtins import range
+from builtins import str
+__author__ = b"dstokes"
+__copyright__ = b"Copyright 2018 Mide Technology Corporation"
 
-__all__ = ['encodeBinary', 'encodeDate', 'encodeFloat', 'encodeId', 'encodeInt',
-           'encodeSize', 'encodeString', 'encodeUInt', 'encodeUnicode']
+__all__ = [b'encodeBinary', b'encodeDate', b'encodeFloat', b'encodeId', b'encodeInt',
```
Comment (StokesMIDE) [inline ebmlite/encoding.py:14]: These probably shouldn't be bytestrings.

**Comment by CFlaniganMide** [comment]
Addressing (almost) all the requested changes at once: I ran everything through an automated converter and manually defined all the string literals as bytes literals to ensure that everything behaved exactly the same in 3.7 as 2.7.  With the exception of the typo, that came when I was swapping StringIO for BytesIO.  I've made the requested changes, looks like everything is running much faster too.

**Comment by CFlaniganMide** [comment]
not actually quite done, got a few more loose ends to tie up

File: ebmlite/decoding.py [hunk 11]
```diff
@@ -15,9 +15,9 @@
 from datetime import datetime, timedelta
 import struct
 
-#===============================================================================
+# ===============================================================================
```
Comment (StokesMIDE) [inline ebmlite/decoding.py:18]: Not sure the space is necessary. In any case, the lines should be shortened.

Comment (CFlaniganMide) [reply ebmlite/decoding.py:19]: Eh, it's technically supposed to have a space after the hash.  It doesn't need it, but my IDE has been complaining.

File: ebmlite/examples.py [hunk 12]
```diff
@@ -30,18 +28,18 @@
                                                         # into a string
 
 # Save xml
-with open(xmlFile1, 'wt') as f:
+with open(xmlFile1, b'wt') as f:
```
Comment (StokesMIDE) [inline ebmlite/examples.py:31]: These bytestring prefixes should be removed from anywhere they aren't necessary (e.g. this and filenames).

File: ebmlite/examples.py [hunk 13]
```diff
@@ -91,11 +89,11 @@
 ebmlDict = ebmlDoc1.dump()
 
 # extract recording properties from EBML
-recProp = ebmlDict['RecordingProperties']
+recProp = ebmlDict[b'RecordingProperties']
```
Comment (StokesMIDE) [inline ebmlite/examples.py:92]: Remove `b` prefix from string keys

File: ebmlite/examples.py [hunk 14]
```diff
@@ -91,11 +89,11 @@
...
 
 # extract recorder info from recording properties
-recInfo = recProp['RecorderInfo']
+recInfo = recProp[b'RecorderInfo']
 
 # extract recorder serial number from recording properties
-recSerial = recInfo['RecorderSerial']
-print('recorder serial number: %d' % recSerial)
+recSerial = recInfo[b'RecorderSerial']
+print(b'recorder serial number: %d' % recSerial)
```
Comment (StokesMIDE) [inline ebmlite/examples.py:99]: Remove `b` prefix from human-read strings.

File: ebmlite/ideParsingExample.py [hunk 15]
```diff
@@ -26,71 +24,71 @@ def getTypeMatch(el, elType):
...
-rawData = b''
+rawData = ''
 payloadType = schema[0xB2]
 for block in dataBlocks:
-    rawData += block.dump()['ChannelDataPayload']
-rawData = np.fromstring(str(rawData), dtype=chEl['ChannelFormat'][1])
-rawData.resize((len(rawData)//3, 3))
+    rawData += block.dump()[b'ChannelDataPayload']
+rawData = np.fromstring(str(rawData), dtype=chEl[b'ChannelFormat'][1])
+rawData.resize((old_div(len(rawData),3), 3))
```
Comment (StokesMIDE) [inline ebmlite/ideParsingExample.py:60]: Can this be done with a regular Python expression? I don't see `old_div()` being imported or defined.

File: tests/test_general.py [hunk 16]
```diff
@@ -142,74 +130,42 @@ def testPPrint(self):
...
 
-        # self.assertTrue(xmlString1 == xmlString2, 'Difference in infinite element, too large to print')
+        # Compare as lists to narrow the location of any differences
+        self.assertListEqual(xmlLines1[1:], xmlLines2[1:],
+                             b'One or more lines are different in the xml documents')
 
 
-if __name__ == "__main__":
-    testsuite = unittest.TestLoader().discover('.')
+if __name__ == b"__main__":
```
Comment (StokesMIDE) [inline tests/test_general.py:169]: More `b` prefixes. Needed?

**Comment by CFlaniganMide** [comment]
I definitely went a bit overboard declaring everything bytes literals, but it got the job done.  Those last two commits should resolve all those issues.

File: tests/test_decode.py [hunk 17]
```diff
@@ -90,11 +90,11 @@ def testReadUInt(self):
                    b'\x15\x41\x42\x43']
         ints = [0x85, 0x4541, 0x254142, 0x15414243]
         
-        for id, i, numOut in zip(idBytes, range(1, len(idBytes) + 1), ints):
-            self.mockStream = StringIO(id)
+        for id, i, numOut in zip(idBytes, list(range(1, len(idBytes) + 1)), ints):
```
Comment (StokesMIDE) [inline tests/test_decode.py:93]: I don't think `range()` needs to be cast to `list`.
Would it be any clearer to implement it like this?
```python
for i, (id, numOut) in enumerate(zip(idBytes, ints), 1):
```

File: tests/test_decode.py [hunk 18]
```diff
@@ -90,11 +90,11 @@ def testReadUInt(self):
...
+            self.mockStream = BytesIO(id)
             self.mockStream.seek(0)
-            val = readUInt(self.mockStream, i)
-            self.assertEquals(val, numOut)
+            a = readUInt(self.mockStream, i)
```
Comment (StokesMIDE) [inline tests/test_decode.py:96]: What was wrong with the variable `val`?

File: tests/test_decode.py [hunk 19]
```diff
@@ -135,74 +135,73 @@ def testReadFloat(self):
...
-        self.assertEqual(val, u'')
+        self.mockStream = BytesIO(b'')
+        self.assertEqual(readString(self.mockStream, 0), b'')
         
-        self.mockStream = StringIO(b'test')
-        val = readString(self.mockStream, 4)
-        self.assertEqual(val, u'test')
+        self.mockStream = BytesIO(b'test')
+        self.assertEqual(readString(self.mockStream, len(self.mockStream.getvalue())),
+                         b'test')
```
Comment (StokesMIDE) [inline tests/test_decode.py:185]: I think using the temporary variable makes the test a little more readable. The first line is a bit too long.

**Pull request closed without merging**